### PR TITLE
MCLOUD-5721: Implement support of new Redis features added in 2.3.5

### PR DESCRIPTION
**Pull request by BaDos**
### Fixed Issues (if relevant)

Added the REDIS_BACKEND environment variable to configure the Redis backend model for Redis cache for Magento 2.3.5 or later.
https://jira.corp.magento.com/browse/MCLOUD-5866

### Manual testing scenarios

I have added FT, but need to recheck these:
1. https://jira.corp.magento.com/browse/MAGECLOUD-17
2. https://jira.corp.magento.com/browse/MAGECLOUD-48

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

releasenotes
Added the REDIS_BACKEND environment variable to configure the Redis backend model for Redis cache for Magento 2.3.5 or later.

### CODE REVIEW

File: src/Config/Factory/Cache.php [hunk 1]
```diff
@@ -183,21 +211,82 @@ private function getSlaveConnection(): array
...
+                    || ($port !== null && $port !== $redisConfig['port'])) {
+                    return false;
+                }
+            }
         }
 
         return true;
     }
+
+    private function getOldConfigStructure(string $envCacheBackendModel, array $redisConfig): array
```
Comment (BarnyShergold) [inline src/Config/Factory/Cache.php]: No ParamBlock

File: src/Config/Factory/Cache.php [hunk 2]
```diff
@@ -183,21 +211,82 @@ private function getSlaveConnection(): array
...
+        return [
+            'backend' => addslashes($envCacheBackendModel),
+            'backend_options' => [
+                'server' => $redisConfig['host'],
+                'port' => $redisConfig['port'],
+            ]
+        ];
+    }
+
+    private function getNewConfigStructure(string $envCacheBackendModel, array $redisConfig): array
```
Comment (BarnyShergold) [inline src/Config/Factory/Cache.php]: No ParamBlock

File: src/Test/Unit/Config/SchemaTest.php [hunk 3]
```diff
@@ -112,6 +112,7 @@ public function testGetDefaultsForDeploy(): void
                 DeployInterface::VAR_LOCK_PROVIDER => 'file',
                 DeployInterface::VAR_CONSUMERS_WAIT_FOR_MAX_MESSAGES => false,
                 DeployInterface::VAR_SPLIT_DB => [],
+                DeployInterface::VAR_CACHE_REDIS_BACKEND => 'Cm_Cache_Backend_Redis'
```
Comment (BarnyShergold) [inline src/Test/Unit/Config/SchemaTest.php]: Missing , at end of line

File: src/Test/Unit/Step/Deploy/PreDeploy/ConfigUpdate/CacheTest.php [hunk 4]
```diff
@@ -120,6 +120,29 @@ public function testExecute()
         $this->step->execute();
     }
 
+    public function executeDataProvider(): array
+    {
+        return [
+            'backend model without remote_backend_options' => [
+                'config' => [
+                    'frontend' => ['frontName' => ['backend' => 'cacheDriver', 'backend_options' => 'some options']]
```
Comment (BarnyShergold) [inline src/Test/Unit/Step/Deploy/PreDeploy/ConfigUpdate/CacheTest.php]: Format better for readability next block is better example

File: src/Test/Unit/Config/Factory/CacheTest.php [hunk 5]
```diff
@@ -149,6 +154,10 @@ public function testGetFromRelationships(
                 [
                     DeployInterface::VAR_REDIS_USE_SLAVE_CONNECTION,
                     $useSlave
```
Comment (BarnyShergold) [inline src/Test/Unit/Config/Factory/CacheTest.php]: missing ,

File: src/Test/Unit/Config/Factory/CacheTest.php [hunk 6]
```diff
@@ -149,6 +154,10 @@ public function testGetFromRelationships(
...
+                ],
+                [
+                    DeployInterface::VAR_CACHE_REDIS_BACKEND,
+                    $backendModel
```
Comment (BarnyShergold) [inline src/Test/Unit/Config/Factory/CacheTest.php]: missing ,

File: src/Test/Unit/Config/Factory/CacheTest.php [hunk 7]
```diff
@@ -234,41 +286,94 @@ public function getFromRelationshipsDataProvider()
...
+            Cache::REDIS_BACKEND_REDIS_CACHE
+        );
+        $resultMasterSlaveConnectionWithDiffHostSyncCache = $resultMasterOnlyConnectionSyncCache;
+        $remoteBackendOptionsDiffHostSync = [
+            'frontend' => [
+                'default' => [
+                    'backend_options' => [
+                        'remote_backend_options' => [
+                            'value' => 'key',
+                            'server' => 'new.host'
```
Comment (BarnyShergold) [inline src/Test/Unit/Config/Factory/CacheTest.php]: Missing ,

File: src/Test/Unit/Config/Factory/CacheTest.php [hunk 8]
```diff
@@ -302,8 +409,158 @@ public function getFromRelationshipsDataProvider()
                 $redisConfiguration,
                 $redisSlaveConfiguration,
                 true,
+                Cache::REDIS_BACKEND_CM_CACHE,
+                5,
                 $resultMasterSlaveConnectionWithDiffHost
             ],
+            [
```
Comment (BarnyShergold) [inline src/Test/Unit/Config/Factory/CacheTest.php:416]: Missing commas on last elements of arrays

**Review by BarnyShergold — CHANGES_REQUESTED** [review]

File: src/Config/Factory/Cache.php [hunk 9]
```diff
@@ -183,21 +211,82 @@ private function getSlaveConnection(): array
...
+                'write_control' => false,
+            ]
+        ];
+    }
+
+    /**
+     * @return bool
+     * @throws \Magento\MagentoCloud\Config\ConfigException
+     */
+    private function isNewConfigStructure(): bool
```
Comment (bbatsche) [inline src/Config/Factory/Cache.php]: "new" and "old" are sort of vague and don't really scale well. Perhaps something more descriptive like "synchronized cache" and...? "unsynced cache"?

**Review by bbatsche — CHANGES_REQUESTED** [review]
I'm not sure what to call the "old" cache configuration ("legacy" is only slightly better) but in general "new"/"old" doesn't seem descriptive enough.

**Review by bbatsche — DISMISSED** [review]

**Comment by BaDos** [comment]
- `src/Test/Functional/Acceptance/PatchApplierCest.php` was removed, it will be added to mageto-cloud-patches
- `src/Test/Functional/Acceptance/PostDeployCest.php` can be ignored, we have the task to fix it https://jira.corp.magento.com/browse/MCLOUD-5968

**Review by bbatsche — DISMISSED** [review]

**Review by bbatsche — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Review by bbatsche — DISMISSED** [review]

**Review by BarnyShergold — DISMISSED** [review]

File: src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php [hunk 10]
```diff
@@ -67,8 +76,19 @@ public function execute()
...
+
+                if (!in_array($backend, CacheFactory::AVAILABLE_REDIS_BACKEND, true)) {
+                    return true;
+                }
+
+                $backendOptions = ($backend === CacheFactory::REDIS_BACKEND_REMOTE_SYNHRONIZED_CACHE)
+                        ? $cacheFrontend['backend_options']['remote_backend_options']
+                        : $cacheFrontend['backend_options'];
+
+                return !in_array($backend, CacheFactory::AVAILABLE_REDIS_BACKEND, true)
```
Comment (oshmyheliuk) [inline src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php]: Looks like you've already done this check `!in_array($backend, CacheFactory::AVAILABLE_REDIS_BACKEND, true)` few lines above, why do you need to do it again?

File: src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php [hunk 11]
```diff
@@ -88,6 +108,30 @@ public function execute()
         $this->configWriter->create($config);
     }
 
+    /**
+     * Checks that configured backend model can be used with installed magento version.
+     *
+     * @param string $backend
+     * @throws StepException
+     * @throws \Magento\MagentoCloud\Package\UndefinedPackageException
```
Comment (oshmyheliuk) [inline src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php]: `StepInterface::execute` can throw only `StepException`

File: src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php (hunk 10 above)
Comment (BaDos) [reply src/Step/Deploy/PreDeploy/ConfigUpdate/Cache.php]: Thanks, I forgot to remove this.

**Review by bbatsche — APPROVED** [review]

**Review by BarnyShergold — APPROVED** [review]

**Comment by mveeramneni** [comment]
QA Approved

**Pull request merged**